pkg/resourcemonitor/noderesourcesaggregator.go

```diff
@@ -0,0 +1,274 @@
+	for nodeID, resList := range perNuma {
+		zone := topologyv1alpha1.Zone{
+			Name:      makeZoneName(nodeID),
+			Type:      "Node",
+			Resources: make(topologyv1alpha1.ResourceInfoList, 0),
+		}
+
+		costs, err := makeCostsPerNumaNode(noderesourceData.topo.Nodes, nodeID)
+		if err != nil {
+			log.Printf("cannot find costs for NUMA node %d: %v", nodeID, err)
+		} else {
+			zone.Costs = topologyv1alpha1.CostList(costs)
+		}
+
+		for name, resData := range resList {
+			allocatableQty := *resource.NewQuantity(resData.allocatable, resource.DecimalSI)
+			capacityQty := *resource.NewQuantity(resData.capacity, resource.DecimalSI)
+			zone.Resources = append(zone.Resources, topologyv1alpha1.ResourceInfo{
+				Name:        name.String(),
+				Allocatable: intstr.FromString(allocatableQty.String()),
+				Capacity:    intstr.FromString(capacityQty.String()),
+			})
+		}
+		zones = append(zones, zone)
+	}
+	return zones
+}
+
+// GetContainerDevicesFromAllocatableResources normalize all compute resources to ContainerDevices.
+// This is helpful because cpuIDs are not represented as ContainerDevices, but with a different format;
+// Having a consistent representation of all the resources as ContainerDevices makes it simpler for
+func GetContainerDevicesFromAllocatableResources(availRes *podresourcesapi.AllocatableResourcesResponse, topo *ghw.TopologyInfo) []*podresourcesapi.ContainerDevices {
+	var contDevs []*podresourcesapi.ContainerDevices
+	for _, dev := range availRes.GetDevices() {
+		contDevs = append(contDevs, dev)
+	}
+
+	cpuIDToNodeIDMap := MakeLogicalCoreIDToNodeIDMap(topo)
+
+	cpusPerNuma := make(map[int][]string)
+	for _, cpuID := range availRes.GetCpuIds() {
+		nodeID, ok := cpuIDToNodeIDMap[int(cpuID)]
+		if !ok {
+			log.Printf("cannot find the NUMA node for CPU %d", cpuID)
+			continue
+		}
+
+		cpuIDList := cpusPerNuma[nodeID]
+		cpuIDList = append(cpuIDList, fmt.Sprintf("%d", cpuID))
+		cpusPerNuma[nodeID] = cpuIDList
+	}
+
+	for nodeID, cpuList := range cpusPerNuma {
+		contDevs = append(contDevs, &podresourcesapi.ContainerDevices{
+			ResourceName: string(v1.ResourceCPU),
+			DeviceIds:    cpuList,
+			Topology: &podresourcesapi.TopologyInfo{
+				Nodes: []*podresourcesapi.NUMANode{
+					{ID: int64(nodeID)},
+				},
+			},
+		})
+	}
+
+	return contDevs
+}
+
+// updateAllocatable computes the actually alloctable resources.
+// This function assumes the allocatable resources are initialized to be equal to the capacity.
+func (noderesourceData *nodeResources) updateAllocatable(numaData map[int]map[v1.ResourceName]*resourceData, ri ResourceInfo) {
+	for _, resID := range ri.Data {
+		resName := string(ri.Name)
+		resMap, ok := noderesourceData.resourceID2NUMAID[resName]
+		if !ok {
+			log.Printf("unknown resource %q", ri.Name)
+			continue
+		}
+		nodeID, ok := resMap[resID]
+		if !ok {
+			log.Printf("unknown resource %q: %q", resName, resID)
+			continue
+		}
+		numaData[nodeID][ri.Name].allocatable--
+	}
```

Comment:
Hmm, are cpus handled somewhere?

Reply:
In the Aggregate method we check for if a certain resource is CPU in order to populate allocatable,capacity and available while accounting for CPUs that are reserved either as part of kube-reserve or system reserve.